File: site/scripts/record-demo.mjs
// Records the landing-page demo from the real Lily TUI and converts it for the player.
//
//   npm run record-demo                  record (python3 + pty) and convert
//   npm run record-demo -- --convert     only re-convert demo/session.cast
//
// 1. scripts/record-demo.py drives `lily --script …` in a pseudo-terminal following
//    demo/storyboard.json and writes demo/session.cast (asciicast v2, raw bytes + timing).
// 2. This script replays the cast through a headless xterm and snapshots the screen after
//    each burst of output. Screens become rows of styled runs, deduplicated into a line
//    table, so a frame is just a list of line ids: demo/demo.json.
import { spawnSync } from "node:child_process";
import { readFileSync, writeFileSync } from "node:fs";
import { homedir, userInfo } from "node:os";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import xterm from "@xterm/headless";

const { Terminal } = xterm;
const SITE = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const STORYBOARD = join(SITE, "demo/storyboard.json");
const CAST = join(SITE, "demo/session.cast");
const OUT = join(SITE, "demo/demo.json");

const FRAME_GAP = 1 / 50; // merge output bursts closer than this (seconds)

// Inverse video on default colours swaps the player's own colours (set in CSS).
const DEFAULT_FG = "var(--term-fg)";
const DEFAULT_BG = "var(--term-bg)";
// ANSI 0–15, tuned to the site's palette; 16–255 are the standard xterm cube and greys.
const BASE16 = ["#2b3326", "#d27a68", "#9fbd87", "#d9c27e", "#8fa9c0", "#bf9bb8", "#8fbcb2", "#d6d6c4", "#6a7462", "#e59a88", "#b7d69f", "#e8d69a", "#a9c1d6", "#d3b2cc", "#a8d3c9", "#f4f2e6"];
function paletteColor(n) {
	if (n < 16) return BASE16[n];
	if (n < 232) {
		const i = n - 16;
		const level = (v) => (v === 0 ? 0 : 55 + v * 40);
		return hex((level(Math.floor(i / 36)) << 16) | (level(Math.floor(i / 6) % 6) << 8) | level(i % 6));
	}
	const g = 8 + (n - 232) * 10;
	return hex((g << 16) | (g << 8) | g);
}
const hex = (n) => `#${n.toString(16).padStart(6, "0")}`;

function record() {
	const result = spawnSync("python3", [join(SITE, "scripts/record-demo.py"), STORYBOARD, CAST], { stdio: "inherit" });
	if (result.error) throw new Error(`could not run python3: ${result.error.message}`);
	if (result.status !== 0) throw new Error(`recording failed (exit ${result.status})`);
}

/** Replaces the local user name and home path (they show up in `ls -la`, paths, …). */
function redactor() {
	const user = (() => {
		try {
			return userInfo().username;
		} catch {
			return "";
		}
	})();
	const home = homedir();
	return (text) => {
		let out = text;
		if (home && home.length > 1) out = out.split(home).join("~");
		if (user && user.length >= 3) {
			// Same width, so `ls -l` columns stay aligned.
			const stand = "lily".padEnd(user.length, " ").slice(0, Math.max(user.length, 4));
			out = out.replace(new RegExp(`\\b${user.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}\\b`, "g"), stand);
		}
		return out;
	};
}

async function convert() {
	const board = JSON.parse(readFileSync(STORYBOARD, "utf8"));
	const [headerLine, ...eventLines] = readFileSync(CAST, "utf8").split("\n").filter(Boolean);
	const header = JSON.parse(headerLine);
	const cols = header.width;
	const rows = header.height;
	const idleLimit = board.idleLimit ?? 1.2;
	const redact = redactor();

	// Output events with idle gaps capped (like asciinema's idle_time_limit).
	const events = [];
	let last = 0;
	let clock = 0;
	for (const line of eventLines) {
		const [t, kind, data] = JSON.parse(line);
		if (kind !== "o") continue;
		clock += Math.min(t - last, idleLimit);
		last = t;
		events.push([clock, data]);
	}
	if (!events.length) throw new Error("the cast has no output");

	const term = new Terminal({ cols, rows, scrollback: 2000, allowProposedApi: true, convertEol: false });
	const write = (data) => new Promise((done) => term.write(data, done));

	const styles = [{}];
	const styleIds = new Map([["{}", 0]]);
	const lines = [];
	const lineIds = new Map();
	const frames = [];
	let cursorVisible = true;
	let lastKey = "";
	const cell = term.buffer.active.getNullCell();

	const colorOf = (value, rgb, palette) => (rgb ? hex(value) : palette ? paletteColor(value) : null);

	const styleId = (c) => {
		let fg = colorOf(c.getFgColor(), c.isFgRGB(), c.isFgPalette());
		let bg = colorOf(c.getBgColor(), c.isBgRGB(), c.isBgPalette());
		if (c.isBold() && c.isFgPalette() && c.getFgColor() < 8) fg = paletteColor(c.getFgColor() + 8);
		if (c.isInverse()) [fg, bg] = [bg ?? DEFAULT_BG, fg ?? DEFAULT_FG];
		const k = [c.isBold() && "b", c.isItalic() && "i", c.isUnderline() && "u", c.isStrikethrough() && "s", c.isDim() && "d", c.isInvisible() && "h"].filter(Boolean).join(" ");
		const css = [fg && `color:${fg}`, bg && `background:${bg}`].filter(Boolean).join(";");
		const style = {};
		if (css) style.c = css;
		if (k) style.k = k;
		const key = JSON.stringify(style);
		let id = styleIds.get(key);
		if (id === undefined) {
			id = styles.length;
			styles.push(style);
			styleIds.set(key, id);
		}
		return id;
	};

	const serializeLine = (y) => {
		const line = term.buffer.active.getLine(y);
		const runs = [];
		if (line) {
			for (let x = 0; x < cols; x++) {
				line.getCell(x, cell);
				if (cell.getWidth() === 0) continue; // second half of a wide character
				const chars = cell.getChars() || " ";
				const id = styleId(cell);
				const lastRun = runs[runs.length - 1];
				if (lastRun && lastRun[1] === id) lastRun[0] += chars;
				else runs.push([chars, id]);
			}
		}
		// Trailing blanks in the default style carry no information.
		while (runs.length) {
			const lastRun = runs[runs.length - 1];
			const style = styles[lastRun[1]];
			if (style.c?.includes("background")) break;
			const trimmed = lastRun[0].replace(/\s+$/, "");
			if (trimmed) {
				lastRun[0] = trimmed;
				break;
			}
			runs.pop();
		}
		for (const run of runs) run[0] = redact(run[0]);
		const key = JSON.stringify(runs);
		let id = lineIds.get(key);
		if (id === undefined) {
			id = lines.length;
			lines.push(runs);
			lineIds.set(key, id);
		}
		return id;
	};

	const snapshot = (t) => {
		if (term.modes.synchronizedOutputMode) return; // mid-frame: the TUI will finish it
		const buffer = term.buffer.active;
		const ids = [];
		for (let y = 0; y < rows; y++) ids.push(serializeLine(buffer.viewportY + y));
		const frame = [Math.round(t * 1000) / 1000, ids, buffer.cursorX, buffer.cursorY, cursorVisible ? 1 : 0];
		const key = JSON.stringify(frame.slice(1));
		if (key === lastKey) return;
		lastKey = key;
		frames.push(frame);
	};

	serializeLine(-1); // line 0 = empty
	for (let i = 0; i < events.length; i++) {
		const [t, data] = events[i];
		for (const m of data.matchAll(/\x1b\[\?25([hl])/g)) cursorVisible = m[1] === "h";
		await write(data);
		const next = events[i + 1];
		if (!next || next[0] - t >= FRAME_GAP) snapshot(t);
	}
	const hold = board.hold ?? 2; // keep the final screen up for a moment before "ended"
	const duration = Math.round((events[events.length - 1][0] + hold) * 1000) / 1000;

	const out = { v: 1, title: board.title ?? header.title ?? "lily", cols, rows, duration, recordedAt: new Date((header.timestamp ?? Date.now() / 1000) * 1000).toISOString(), styles, lines, frames };
	writeFileSync(OUT, `${JSON.stringify(out)}\n`);
	const kb = (Buffer.byteLength(JSON.stringify(out)) / 1024).toFixed(0);
	console.log(`record-demo: wrote demo/demo.json — ${frames.length} frames, ${lines.length} unique lines, ${styles.length} styles, ${duration.toFixed(1)}s, ${kb} KB`);
}

const convertOnly = process.argv.includes("--convert");
try {
	if (!convertOnly) record();
	await convert();
	console.log("record-demo: run `npm run build` to publish it into dist/");
} catch (error) {
	console.error(`record-demo: ${error.message}`);
	process.exit(1);
}

File: site/scripts/record-demo.py
"""Records a real Lily TUI session in a pseudo-terminal as an asciicast v2 file.

    python3 scripts/record-demo.py demo/storyboard.json demo/session.cast

Usually run through `npm run record-demo`, which also converts the cast into the frames the
landing page replays (scripts/record-demo.mjs). The storyboard lists what to type and what
to wait for; the driver behaves like tui_smoke.py in the repo: it answers the terminal's
DA1 query, types with human-ish delays and pauses before Enter.
"""
import codecs
import fcntl
import json
import os
import pty
import random
import re
import select
import shutil
import signal
import struct
import sys
import termios
import time

HERE = os.path.dirname(os.path.abspath(__file__))
SITE = os.path.dirname(HERE)

storyboard_path = sys.argv[1] if len(sys.argv) > 1 else os.path.join(SITE, "demo", "storyboard.json")
cast_path = sys.argv[2] if len(sys.argv) > 2 else os.path.join(SITE, "demo", "session.cast")
board = json.load(open(storyboard_path))
site_path = lambda p: p if os.path.isabs(p) else os.path.normpath(os.path.join(SITE, p))

cols, rows = int(board.get("cols", 92)), int(board.get("rows", 26))
lily = site_path(board.get("lily", "../bin/lily.mjs"))
script = site_path(board.get("script", "../examples/scripts/chat.json"))
workspace_src = site_path(board.get("workspace", "demo/workspace"))
workspace = board.get("workspaceDir", "/tmp/lily-demo/tidy-notes")
home = board.get("home", "/tmp/lily-demo/home")
backend = os.environ.get("LILY_DEMO_BACKEND") or board.get("backend")
random.seed(board.get("seed", 7))

# A fresh workspace (a copy of the small demo repo) and a fresh Lily home on every recording.
for path in (workspace, home):
    shutil.rmtree(path, ignore_errors=True)
shutil.copytree(workspace_src, workspace)
os.makedirs(home, exist_ok=True)

argv = ["node", lily, "--script", script] + (["--backend", backend] if backend else [])
pid, fd = pty.fork()
if pid == 0:
    fcntl.ioctl(0, termios.TIOCSWINSZ, struct.pack("HHHH", rows, cols, 0, 0))
    os.chdir(workspace)
    env = dict(os.environ, LILY_HOME=home, TERM="xterm-256color", COLORTERM="truecolor", COLUMNS=str(cols), LINES=str(rows))
    env.pop("NO_COLOR", None)
    os.execvpe("node", argv, env)

start = time.time()
events = []
recording = True
decoder = codecs.getincrementaldecoder("utf-8")("replace")
screen_text = ""  # everything printed so far, escape sequences stripped (for waitFor)
last_output = time.time()
ANSI = re.compile(r"\x1b\[[0-9;?<>=]*[ -/]*[@-~]|\x1b\][^\x07\x1b]*(?:\x07|\x1b\\)|\x1b[_P^][^\x1b]*\x1b\\|\x1b[()][0-9A-Za-z]|\x1b[=>78DEHMNOZc]")


def stamp():
    return round(time.time() - start, 4)


def pump(seconds):
    """Reads output for `seconds`, answering the queries a real terminal would answer."""
    global screen_text, last_output
    end = time.time() + seconds
    while True:
        remaining = end - time.time()
        if remaining <= 0:
            return True
        r, _, _ = select.select([fd], [], [], min(remaining, 0.05))
        if not r:
            continue
        try:
            chunk = os.read(fd, 65536)
        except OSError:
            return False
        if not chunk:
            return False
        if b"\x1b[c" in chunk or b"\x1b[0c" in chunk:
            os.write(fd, b"\x1b[?62;22c")  # DA1: a VT220-ish terminal with ANSI colour
        text = decoder.decode(chunk)
        if text:
            last_output = time.time()
            screen_text = (screen_text + ANSI.sub("", text))[-200000:]
            if recording:
                events.append([stamp(), "o", text])


def send(data):
    if recording:
        events.append([stamp(), "i", data])
    os.write(fd, data.encode())


def wait_for(pattern, timeout):
    end = time.time() + timeout
    offset = len(screen_text)
    while time.time() < end:
        pump(0.1)
        if re.search(pattern, screen_text[max(0, offset - 4000):]) or re.search(pattern, screen_text):
            return True
    return False


def wait_idle(quiet, timeout):
    """Waits until the TUI printed nothing for `quiet` seconds (bounded by `timeout`)."""
    end = time.time() + timeout
    pump(0.2)
    while time.time() < end:
        pump(0.1)
        if time.time() - last_output >= quiet:
            return True
    return False


KEYS = {"enter": "\r", "escape": "\x1b", "tab": "\t", "up": "\x1b[A", "down": "\x1b[B", "ctrl-c": "\x03", "backspace": "\x7f"}

for step in board["steps"]:
    if "waitFor" in step:
        if not wait_for(step["waitFor"], step.get("timeout", 20)):
            print(f"record-demo: timed out waiting for {step['waitFor']!r}", file=sys.stderr)
            print(screen_text[-3000:], file=sys.stderr)
            os.kill(pid, signal.SIGKILL)
            sys.exit(1)
    elif "idle" in step:
        wait_idle(step["idle"], step.get("timeout", 30))
    elif "pause" in step:
        pump(step["pause"])
    elif "type" in step:
        for ch in step["type"]:
            send(ch)
            base = step.get("delay", 0.075)
            pump(base * (0.6 + random.random() * 0.9) + (0.12 if ch == " " and random.random() < 0.25 else 0))
    elif "key" in step:
        send(KEYS[step["key"]])
        pump(0.05)
    elif step.get("end"):
        recording = False
        break

# Leave the TUI the way a person would: Ctrl+C twice.
send("\x03")
pump(0.4)
send("\x03")
pump(1.5)
try:
    os.kill(pid, signal.SIGTERM)
except ProcessLookupError:
    pass
try:
    os.waitpid(pid, 0)
except ChildProcessError:
    pass

os.makedirs(os.path.dirname(cast_path), exist_ok=True)


def redact(text):
    """Keeps the local user name and home path out of the checked-in cast (same width, so
    `ls -l` columns stay aligned)."""
    home_dir = os.path.expanduser("~")
    if len(home_dir) > 1:
        text = text.replace(home_dir, "~")
    user = os.environ.get("USER") or os.environ.get("LOGNAME") or ""
    if len(user) >= 3:
        stand = "lily".ljust(len(user))[: max(len(user), 4)]
        text = re.sub(rf"\b{re.escape(user)}\b", stand, text)
    return text


with open(cast_path, "w") as out:
    header = {
        "version": 2,
        "width": cols,
        "height": rows,
        "timestamp": int(start),
        "title": board.get("title", "lily"),
        "env": {"TERM": "xterm-256color"},
    }
    out.write(json.dumps(header) + "\n")
    for event in events:
        out.write(json.dumps([event[0], event[1], redact(event[2])], ensure_ascii=False) + "\n")
if not os.environ.get("LILY_DEMO_KEEP"):
    for path in (workspace, home):
        shutil.rmtree(path, ignore_errors=True)
print(f"record-demo: wrote {cast_path} ({len(events)} events, {events[-1][0] if events else 0:.1f}s)")
